Add experimental support for tracking the invocations of runtime functions

swiftix opened the pull request:
It allows for collecting the state of runtime function counters, which are used to determine how many times a given runtime function was called.

It is possible to get the global counters, which represent the total number of invocations, or per-object counters, which represent the number of runtime functions calls for a specific object.

The APIs are implemented in the runtime library and are exposed as Swift APIs via the standard library. This way it is possible to write tests and benchmarks which collect very precise data about the amount of runtime functions invocations (e.g. swift_retain/swift_release) for an object or for a given code snippet.

swiftix committed:
Add runtime library support for tracking the invocations of runtime functions.

swiftix committed:
Add experimental Swift standard library support for tracking the invocations of runtime functions

It allows for collecting the state of runtime function counters, which are used to determine how many times a given runtime function was called.

It is possible to get the global counters, which represent the total number of invocations, or per-object counters, which represent the number of runtime functions calls for a specific object.

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@moiseev Max, could you have a look at the stdlib part?

swiftix commented:
@aschwaighofer Oh, I see, you are right!

swiftix committed:
Add a test for testing the runtime function counters functionality

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test

swiftix commented:
@gottesmm Michael, do you want to have a look at this attempt to provide runtime function counters?

swiftix committed:
Fix creation of UnsafeRawPointer for found references

swiftix commented:
@swift-ci please smoke test OS X

swiftix commented:
@swift-ci please smoke test OS X

swiftix commented:
@swift-ci please test OS X

swiftix commented:
@swift-ci please test OS X

swiftix commented:
@swift-ci please test OS X

swiftix commented:
@swift-ci please test OS X

swiftix commented:
@swift-ci please test OS X

swiftix commented:
@swift-ci please smoke test linux

swiftix committed:
Provide runtime function counters for object allocation, deallocation and initialization of static or stack-promoted objects

swiftix commented:
@swift-ci please smoke test

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+
+  // Recursively visit the children of the current value.
+  var currentIndex = mirror.children.startIndex
+  for _ in 0..<count {
```
Why can't you just `for child in mirror.children`?

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  }
+}
+
+/// This is a namespace for runtime functions related to management
```
`///` is used for doc comments only.

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+    _RuntimeFunctionCounters.getRuntimeFunctionCountersOffsets()
+  public static let numRuntimeFunctionCounters =
+    _RuntimeFunctionCounters.getNumRuntimeFunctionCounters()
+  public static let runtimtFunctionNameToIndex: [String : Int] =
```
`s/runtimt/runtime`

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+      _RuntimeFunctionCounters.getNumRuntimeFunctionCounters()
+    var runtimtFunctionNames : [String] = []
+    for index in 0..<numRuntimeFunctionCounters {
+      let name = String(cString:runtimeFunctionNames[index])
```
Just a nitpick: space after the `:`.

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+    let runtimeFunctionNames = _RuntimeFunctionCounters._getRuntimeFunctionNames()
+    let numRuntimeFunctionCounters =
+      _RuntimeFunctionCounters.getNumRuntimeFunctionCounters()
+    var runtimtFunctionNames : [String] = []
```
Since you already know the future length of a resulting array, calling `runtimeFunctionNames.reserveCapacity(numRuntimeFunctionCounters)` will help avoid reallocation.

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  @_silgen_name("setGlobalRuntimeFunctionCountersUpdateHandler")
+  static public func setGlobalRuntimeFunctionCountersUpdateHandler(
+    handler: RuntimeFunctionCountersUpdateHandler?) ->
+    RuntimeFunctionCountersUpdateHandler?
```
I started to prefer the formatting where `->` is on the same line with the function result type. As in:

```swift
static public func setGlobalRuntimeFunctionCountersUpdateHandler(
  handler: RuntimeFunctionCountersUpdateHandler?
) -> RuntimeFunctionCountersUpdateHandler?
```

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+    [UnsafeRawPointer] {
+    var visited : [ObjectIdentifier : Int] = [:]
+    var references: [UnsafeRawPointer] = []
+    _collectAllReferencesInsideObjectImpl(value,
```
```swift
_collectAllReferencesInsideObjectImpl(
  value, references: &references, visitedItems: &visited)
```

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  static internal func getRuntimeFunctionNameToIndex() -> [String : Int] {
+    let runtimeFunctionNames = _RuntimeFunctionCounters.getRuntimeFunctionNames()
+    var runtimtFunctionNameToIndex : [String : Int] = [:]
+    let numRuntimeFunctionCounters =
```
Same trick with calling `reserveCapacity` applies to dictionaries.

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  init()
+
+  /// Dump the current state of all counters.
+  func dump(skipUnchanged: Bool)
```
How about this?

```swift
protocol _RuntimeFunctionCountersStats {
  func dump<T : TextOutputStream>(skipUnchanged: Bool, to: inout T)
  func dumpDiff<T : TextOutputStream>(_ after: Self, skipUnchanged: Bool, to: inout T)
}
```

And then:

```swift
extension _RuntimeFunctionCounterStats {
  public func dump(skipUnchanged: Bool) {
    dump(skipUnchanged: skipUnchanged, to: &_StdOut)
  }
}

extension _RuntimeFunctionCountersStats : CustomDebugStringConvertible {
  public var debugDescription: String {
    var result = ""
    dump(skipUnchanged: true, to: &result)
    return result
  }
}
```

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  )
+
+  // Use counter name as index.
+  subscript(_ index: String) -> UInt32 {
```
`index` is misleading here. Just call it `counterName` =)

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+  public // @testable
+  func enableRuntimeFunctionCountersUpdates(
+    mode: (globalMode: Bool, perObjectMode: Bool) = (true, true)) {
+    _ = _RuntimeFunctionCounters.setGlobalRuntimeFunctionCountersMode(
```
`_ =` should become unnecessary once the function is marked with a `@discardableResult` attribute.

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+/// counters related to a given object.
+public // @testable
+func _measureRuntimeFunctionCountersDiffs(
+  objects: [UnsafeRawPointer], _ f: () -> Void) ->
```
`f` is not very descriptive. `body` perhaps? At least it has some part of the overall meaning...

moiseev commented on stdlib/public/core/RuntimeFunctionCounters.swift:
```diff
@@ -0,0 +1,506 @@
+    f()
+    // Disable counters updates.
+    _RuntimeFunctionCounters.enableRuntimeFunctionCountersUpdates(
+    mode: (globalMode: false, perObjectMode: false))
```
Indent this line please.

swiftix committed:
Address review comments related to the Swift standard library part of the PR

swiftix commented:
@moiseev Max, thanks a lot for your review! I addressed all of your comments and pushed a new version. Waiting for @gottesmm to review the C++ part of the PR.

swiftix commented:
@swift-ci please smoke test

gparker42 commented:
I don't see any thread-safety here. Are these counters supposed to be atomic?

swiftix commented:
@gparker42 I've put into the comments that we may want to make those counters atomic. But in principle, the main motivation for them is to write some tests and perform some performance investigations to track e.g. the number of ARC operations being performed during the execution of a given stdlib call like `Array.append`,`Dictionary.get` or something like this. For this kind of use-cases, we don't need to be thread-safe IMHO. WDYT?

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.cpp:
```diff
@@ -0,0 +1,215 @@
+static llvm::DenseMap<HeapObject *, RuntimeFunctionCountersState> RuntimeObjectStateCache;
+
+/// Define names of runtime functions.
+#undef RT_FUNCTION_TO_TRACK
```
Why are you undefing this? This would be much simpler with a .def file.

gottesmm reviewed (commented):
A bunch of comments. Most of it comes down to is this a c++ or a c api and also can you use a .def file? The way you are using macros here is very unstructured, using a .def will be cleaner.

swiftix committed:
Address review comments.

swiftix commented:
@gottesmm Thanks for the review, Michael! I pushed a new version where I addressed most of your comments.

swiftix commented:
@swift-ci please smoke test

swiftix committed:
Fix a macro-definition

swiftix commented:
@swift-ci please smoke test

swiftix committed:
The test needs the standard library (the runtime library part of it) built with assertions

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@swift-ci please smoke test

swiftix commented:
@gottesmm Michael, is there anything still missing before I merge it? 
The atomicity handling can be  added later if we'd really need it (see my comments above).

gottesmm commented:
I am re-reading now.

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.h:
```diff
@@ -0,0 +1,110 @@
+} // end namespace swift
+
+/// The name of a helper function for tracking the calls of a runtime function.
+#define SWIFT_RT_TRACK_INVOCATION_NAME(RT_FUNCTION)                            \
```
Why is this outside the #if !defined(NDEBUG)

swiftix replied:
Because it is used also by the RuntimeInvocationsTracking.cpp, which is compiled always, even if NDEBUG is defined.

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.h:
```diff
@@ -0,0 +1,162 @@
+/// Get the runtime object state associated with an object and store it
+/// into the result.
+extern "C" void
+getObjectRuntimeFunctionCounters(swift::HeapObject *object,
```
This is better, but you are still using C++ namespaces. If this is intended to be a C API, these should all be typedefs of void * or of an anonymous C struct.

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.def:
```diff
@@ -0,0 +1,55 @@
+/// FUNCTION_TO_TRACK(Id)
+/// Id is the name of the runtime function to be tracked.
+
+/// Set of runtime functions that whose invocations need to be tracked.
```
This makes all of the code look much nicer! Looks really nice! One more suggestion: there is no reason why one should include this if one doesn't define FUNCTION_TO_TRACK. I would add:

```
#ifndef FUNCTION_TO_TRACK
#error "Must define FUNCTION_TO_TRACK to include RuntimeInvocationsTracking.def"
#endif
```

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.cpp:
```diff
@@ -0,0 +1,231 @@
+}
+
+/// Define how to dump the counter for a given runtime function.
+#define FUNCTION_TO_TRACK(RT_FUNCTION)                                         \
```
We generally put this /right/ next to the `#include`.

So I would put it like so:

```
static void dumpRuntimeCounters(...) {
uint32_t tmp = 0;
#define FUNCTION_TO_TRACK(RT_FUNCTION) \
   ... \
   ...
#include "RuntimeInvocationsTracking.def"
}
```

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.cpp:
```diff
@@ -0,0 +1,231 @@
+/// Return the names of the runtime functions being tracked.
+/// Their order is the same as the order of the counters in the
+/// RuntimeObjectState structure.
+SWIFT_RT_ENTRY_VISIBILITY
```
Are these null terminated? If so, you should document that.

gottesmm commented on stdlib/public/runtime/RuntimeInvocationsTracking.cpp:
```diff
@@ -0,0 +1,231 @@
+    GlobalRuntimeFunctionCountersUpdateHandler;
+
+/// Define offset for each function being tracked.
+#define FUNCTION_TO_TRACK(RT_FUNCTION)                                         \
```
Can you move this next to the include of RuntimeInvocationsTracking.def in the function itself.

gottesmm reviewed (commented):
It looks better. We generally put the `#define` next to the `#include` when using XMACRO lists. I pointed out a few areas where this happens. Can you fix those and the rest?

swiftix committed:
Expose runtime APIs for runtime function counters in such a way that they can be used from C

swiftix committed:
Warn if FUNCTION_TO_TRACK is not defined

swiftix committed:
Move FUNCTION_TO_TRACK definitions closer to the includes of "RuntimeInvocationsTracking.def"

swiftix committed:
#include should be on the next line after #define

swiftix committed:
Add a comment explaining why the file needs to be compiled even in no-asserts builds

swiftix committed:
Check that the runtime counters APIs are available even in no-asserts builds

No runtime function calls will be tracked in this case, but the APIs for reading them are still available. This way there is no need to recompile Swift clients if they are linked against different builds of the standard library and runtime library.

swiftix commented:
@swift-ci please smoke test

gottesmm commented:
LGTM

swiftix commented:
Big thanks to @gottesmm for his thorough review!

swiftix commented:
@swift-ci please test

swiftix commented:
@swift-ci please test
